server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -376,32 +493,63 @@ remoteField
   -> Maybe G.Description
   -> G.ArgumentsDefinition
   -> G.TypeDefinition [G.Name]
-  -> m (FieldParser n ()) -- TODO return something useful, maybe?
+  -> m (FieldParser n (Field NoFragments G.Name))
 remoteField sdoc fieldName description argsDefn typeDefn = do
   -- TODO add directives
   argsParser <- argumentsParser argsDefn sdoc
   case typeDefn of
     G.TypeDefinitionObject objTypeDefn -> do
-      remoteSchemaObj <- remoteSchemaObject sdoc objTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
-    G.TypeDefinitionScalar (G.ScalarTypeDefinition desc name' _) ->
-      pure $ P.selection fieldName description argsParser $ remoteFieldScalarParser name' desc
+      remoteSchemaObject sdoc objTypeDefn >>=
+        mkFieldParserWithSelectionSet fieldName description argsParser
+    G.TypeDefinitionScalar scalarTypeDefn ->
+      let scalarField = remoteFieldScalarParser scalarTypeDefn
+      in mkFieldParserWithoutSelectionSet fieldName description argsParser scalarField
     G.TypeDefinitionEnum enumTypeDefn ->
-      pure $ P.selection fieldName description argsParser $ remoteFieldEnumParser enumTypeDefn
-    G.TypeDefinitionInterface ifaceTypeDefn -> do
-      remoteSchemaObj <- remoteSchemaInterface sdoc ifaceTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
-    G.TypeDefinitionUnion unionTypeDefn -> do
-      remoteSchemaObj <- remoteSchemaUnion sdoc unionTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
+      let enumField = remoteFieldEnumParser enumTypeDefn
+      in mkFieldParserWithoutSelectionSet fieldName description argsParser enumField
+    G.TypeDefinitionInterface ifaceTypeDefn ->
+      remoteSchemaInterface sdoc ifaceTypeDefn >>=
+        mkFieldParserWithSelectionSet fieldName description argsParser
+    G.TypeDefinitionUnion unionTypeDefn ->
+      remoteSchemaUnion sdoc unionTypeDefn >>=
+        mkFieldParserWithSelectionSet fieldName description argsParser
     _ -> throw400 RemoteSchemaError "expected output type, but got input type"
+  where
+    mkFieldParserWithoutSelectionSet
+      :: G.Name
+      -> (Maybe Description)
+      -> (InputFieldsParser n ())
+      -> (Parser 'Both n ())
+      -> m (FieldParser n (Field NoFragments G.Name))
+    mkFieldParserWithoutSelectionSet fldName desc argsParser outputParser =
+      let selectionFldParser = P.rawSelection fldName desc argsParser outputParser
+      in
+        pure $
+          (selectionFldParser
+           <&>
+           (\(alias, args, _)
+              -> (G.Field alias fieldName (fmap getName <$> args) mempty [])))
```

Comment:
Since this is always `pure`, rather make the function `mkFieldParserWithoutSelectionSet` itself pure and use `<&>` rather than `>>=` above in the 2 cases.

Reply:
The `let ... in ...` syntax doesn't add much to the readability for me. If you want to clarify why you're using `P.rawSelection`, you can write inline comments, even in the middle of an expression.